pkg/pubsub/adapter/adapter.go

```diff
@@ -18,253 +18,246 @@ package adapter
-		if a.transformer == nil {
-			if a.transformer, err = kncloudevents.NewDefaultClient(a.Transformer); err != nil {
-				return fmt.Errorf("failed to create transformer cloudevent client: %w", err)
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				a.logger.Warn("Failed to close response body", zap.Error(err))
 			}
-		}
-	}
+		}()
 
-	return a.inbound.StartReceiver(ctx, a.receive)
-}
+		a.reporter.ReportEventCount(args, resp.StatusCode)
 
-func (a *Adapter) receive(ctx context.Context, event cloudevents.Event, resp *cloudevents.EventResponse) error {
-	logger := logging.FromContext(ctx).With(zap.Any("event.id", event.ID()), zap.Any("sink", a.Sink))
+		if resp.StatusCode/100 != 2 {
+			a.logger.Error("Event delivery failed", zap.Int("StatusCode", resp.StatusCode))
+			msg.Nack()
+			return
+		}
 
-	// TODO Name and ResourceGroup might cause problems in the near future, as we might use a single receive-adapter
-	//  for multiple source objects. Same with Namespace, when doing multi-tenancy.
-	args := &ReportArgs{
-		Name:          a.Name,
-		Namespace:     a.Namespace,
-		EventType:     event.Type(),
-		EventSource:   event.Source(),
-		ResourceGroup: a.ResourceGroup,
-	}
+		respMsg := cehttp.NewMessageFromHttpResponse(resp)
+		if respMsg.ReadEncoding() == binding.EncodingUnknown {
+			// No reply
+			msg.Ack()
+			return
+		}
 
-	var err error
-	// If a transformer has been configured, then transform the message.
-	// Note that this path in the code will be executed when using the receive adapter as part of the underlying Channel
-	// of a Broker. We currently set the TransformerURI to be the address of the Broker filter pod.
-	// TODO consider renaming transformer as it is confusing.
-	if a.transformer != nil {
-		transformedCTX, transformedEvent, err := a.transformer.Send(ctx, event)
-		rtctx := cloudevents.HTTPTransportContextFrom(transformedCTX)
+		// If there was a reply, we need to send it to the sink.
+		// We then overwrite the initial event we sent.
+		event, err = binding.ToEvent(ctx, respMsg)
 		if err != nil {
-			logger.Errorf("error transforming cloud event %q", event.ID())
-			a.reporter.ReportEventCount(args, rtctx.StatusCode)
-			return err
+			a.logger.Error("Failed to convert response message to event",
+				zap.Any("response", respMsg), zap.Error(err))
+			msg.Ack()
+			return
 		}
-		if transformedEvent == nil {
-			// This doesn't mean there was an error. E.g., the Broker filter pod might not return a response.
-			// Report the returned Status Code and return.
-			logger.Debugf("cloud event %q was not transformed", event.ID())
-			a.reporter.ReportEventCount(args, rtctx.StatusCode)
-			return nil
-		}
-		// Update the event with the transformed one.
-		event = *transformedEvent
+
 		// Update the tracing information to use the span returned by the transformer.
-		ctx = trace.NewContext(ctx, trace.FromContext(transformedCTX))
+		var span *trace.Span
+		ctx, span = a.startSpan(ctx, event)
```

Comment:
I don't think this is right place to start a new span. We should instead start a single span at the start of receive which will be the parent of both the transform and outbound transactions. The span name should probably be derived from the name of the source in this case.

Reply:
updated it. Thanks for the feedback on this one.
Let's see if tests pass